File: backend/engine/achievements.py
"""
GridForge Achievements — Python port of src/engine/Achievements.js

Achievement definitions, stat building, and checking logic.
"""

from __future__ import annotations

from .constants import MIN_SOC, MAX_SOC


ACHIEVEMENTS = [
    # BM mastery
    {"id": "FIRST_CLEAR", "name": "First Blood", "desc": "Get your first bid accepted in the BM",
     "check": lambda s: s.get("totalAccepted", 0) >= 1},
    {"id": "STREAK_5", "name": "Hot Streak", "desc": "5 consecutive accepted bids",
     "check": lambda s: s.get("streak", 0) >= 5},
    {"id": "STREAK_10", "name": "Unstoppable", "desc": "10 consecutive accepted bids",
     "check": lambda s: s.get("streak", 0) >= 10},
    # Revenue milestones
    {"id": "EARN_500", "name": "Profitable Trader", "desc": "Earn £500 total revenue",
     "check": lambda s: s.get("totalRevenue", 0) >= 500},
    {"id": "EARN_2000", "name": "Market Maker", "desc": "Earn £2,000 total revenue",
     "check": lambda s: s.get("totalRevenue", 0) >= 2000},
    {"id": "EARN_5000", "name": "Whale", "desc": "Earn £5,000 total revenue",
     "check": lambda s: s.get("totalRevenue", 0) >= 5000},
    # Asset-specific
    {"id": "BATTERY_MASTER", "name": "Battery Master", "desc": "Operate a battery through 10+ SPs without hitting SoC limits",
     "check": lambda s: s.get("assetKind") == "soc" and s.get("totalSPs", 0) >= 10 and not s.get("hitSoCLimit")},
    {"id": "WIND_WHISPERER", "name": "Wind Whisperer", "desc": "Earn £1,000+ with a wind farm",
     "check": lambda s: s.get("assetKey") == "WIND" and s.get("totalRevenue", 0) >= 1000},
    {"id": "GAS_KING", "name": "Scarcity Shark", "desc": "Earn £500+ in a single SP with OCGT",
     "check": lambda s: s.get("assetKey") == "OCGT" and s.get("bestSingleSP", 0) >= 500},
    {"id": "FLEX_LORD", "name": "Flex Lord", "desc": "Accept 15+ bids with DSR",
     "check": lambda s: s.get("assetKey") == "DSR" and s.get("totalAccepted", 0) >= 15},
    # Scenario-specific
    {"id": "SURVIVE_DUNKEL", "name": "Dark Survivor", "desc": "Stay profitable during Dunkelflaute",
     "check": lambda s: s.get("scenario") == "DUNKELFLAUTE" and s.get("totalRevenue", 0) > 0},
    {"id": "SPIKE_RIDER", "name": "Spike Rider", "desc": "Earn £1,000+ during a Scarcity Event scenario",
     "check": lambda s: s.get("scenario") == "SPIKE" and s.get("totalRevenue", 0) >= 1000},
    # DA market
    {"id": "DA_WINNER", "name": "Forward Thinker", "desc": "Earn £500+ from Day-Ahead auctions",
     "check": lambda s: s.get("daCash", 0) >= 500},
    # Strategic
    {"id": "PERFECT_TIMING", "name": "Perfect Timing", "desc": "Buy low and sell high within 3 SPs",
     "check": lambda s: s.get("hadBuySellFlip", False)},
    {"id": "SURVIVOR", "name": "Grid Guardian", "desc": "Play 20+ SPs without triggering frequency breach",
     "check": lambda s: s.get("totalSPs", 0) >= 20 and not s.get("hadFreqBreach")},
]


def build_achievement_stats(data: dict) -> dict:
    sp_history = data.get("spHistory", [])
    cash = data.get("cash", 0)
    da_cash = data.get("daCash", 0)
    asset_key = data.get("assetKey", "")
    asset_kind = data.get("assetKind", "")
    scenario = data.get("scenario", "")
    soc = data.get("soc", 50)
    freq_breach_sec = data.get("freqBreachSec", 0)

    total_sps = len(sp_history)
    accepted = [h for h in sp_history if h.get("accepted")]
    total_accepted = len(accepted)
    total_revenue = cash
    best_single_sp = max((h.get("revenue", 0) for h in accepted), default=0)

    streak = 0
    for h in reversed(sp_history):
        if h.get("accepted"):
            streak += 1
        else:
            break

    hit_soc_limit = soc <= (MIN_SOC + 1) or soc >= (MAX_SOC - 1)

    had_buy_sell_flip = (
        len(sp_history) >= 3
        and any(h.get("accepted") and not h.get("isShort") for h in sp_history[-3:])
        and any(h.get("accepted") and h.get("isShort") for h in sp_history[-3:])
    )

    return {
        "totalSPs": total_sps,
        "totalAccepted": total_accepted,
        "totalRevenue": total_revenue,
        "bestSingleSP": best_single_sp,
        "streak": streak,
        "assetKey": asset_key,
        "assetKind": asset_kind,
        "scenario": scenario,
        "daCash": da_cash,
        "soc": soc,
        "hitSoCLimit": hit_soc_limit,
        "hadBuySellFlip": had_buy_sell_flip,
        "hadFreqBreach": freq_breach_sec > 0,
    }


def check_achievements(stats: dict, already_earned: list[str]) -> list[dict]:
    newly_earned = []
    for a in ACHIEVEMENTS:
        if a["id"] in already_earned:
            continue
        try:
            if a["check"](stats):
                newly_earned.append(a)
        except Exception as exc:
            import logging
            logging.getLogger(__name__).warning(
                "Achievement check %s failed: %s", a["id"], exc
            )
    return newly_earned

File: frontend/src/engine/Achievements.js
import { MIN_SOC, MAX_SOC } from '../shared/constants.js';

export const ACHIEVEMENTS = [
    // BM mastery
    {
        id: "FIRST_CLEAR", name: "First Blood", emoji: "🎯", desc: "Get your first bid accepted in the BM", col: "#1de98b",
        check: (s) => s.totalAccepted >= 1
    },
    {
        id: "STREAK_5", name: "Hot Streak", emoji: "🔥", desc: "5 consecutive accepted bids", col: "#f5b222",
        check: (s) => s.streak >= 5
    },
    {
        id: "STREAK_10", name: "Unstoppable", emoji: "💎", desc: "10 consecutive accepted bids", col: "#b78bfa",
        check: (s) => s.streak >= 10
    },

    // Revenue milestones
    {
        id: "EARN_500", name: "Profitable Trader", emoji: "💰", desc: "Earn £500 total revenue", col: "#1de98b",
        check: (s) => s.totalRevenue >= 500
    },
    {
        id: "EARN_2000", name: "Market Maker", emoji: "🏦", desc: "Earn £2,000 total revenue", col: "#38c0fc",
        check: (s) => s.totalRevenue >= 2000
    },
    {
        id: "EARN_5000", name: "Whale", emoji: "🐋", desc: "Earn £5,000 total revenue", col: "#b78bfa",
        check: (s) => s.totalRevenue >= 5000
    },

    // Asset-specific
    {
        id: "BATTERY_MASTER", name: "Battery Master", emoji: "🔋", desc: "Operate a battery through 10+ SPs without hitting SoC limits", col: "#1de98b",
        check: (s) => s.assetKind === "soc" && s.totalSPs >= 10 && !s.hitSoCLimit
    },
    {
        id: "WIND_WHISPERER", name: "Wind Whisperer", emoji: "🌬️", desc: "Earn £1,000+ with a wind farm", col: "#a3e635",
        check: (s) => s.assetKey === "WIND" && s.totalRevenue >= 1000
    },
    {
        id: "GAS_KING", name: "Scarcity Shark", emoji: "🦈", desc: "Earn £500+ in a single SP with OCGT", col: "#f0455a",
        check: (s) => s.assetKey === "OCGT" && s.bestSingleSP >= 500
    },
    {
        id: "FLEX_LORD", name: "Flex Lord", emoji: "🏗️", desc: "Accept 15+ bids with DSR", col: "#f5b222",
        check: (s) => s.assetKey === "DSR" && s.totalAccepted >= 15
    },

    // Scenario-specific
    {
        id: "SURVIVE_DUNKEL", name: "Dark Survivor", emoji: "🌑", desc: "Stay profitable during Dunkelflaute", col: "#f0455a",
        check: (s) => s.scenario === "DUNKELFLAUTE" && s.totalRevenue > 0
    },
    {
        id: "SPIKE_RIDER", name: "Spike Rider", emoji: "🚀", desc: "Earn £1,000+ during a Scarcity Event scenario", col: "#f5b222",
        check: (s) => s.scenario === "SPIKE" && s.totalRevenue >= 1000
    },

    // DA market
    {
        id: "DA_WINNER", name: "Forward Thinker", emoji: "📋", desc: "Earn £500+ from Day-Ahead auctions", col: "#f5b222",
        check: (s) => s.daCash >= 500
    },

    // Strategic
    {
        id: "PERFECT_TIMING", name: "Perfect Timing", emoji: "⏱️", desc: "Buy low and sell high within 3 SPs", col: "#38c0fc",
        check: (s) => s.hadBuySellFlip
    },
    {
        id: "SURVIVOR", name: "Grid Guardian", emoji: "🛡️", desc: "Play 20+ SPs without triggering frequency breach", col: "#67e8f9",
        check: (s) => s.totalSPs >= 20 && !s.hadFreqBreach
    },
];

export function buildAchievementStats({ spHistory, cash, daCash, assetKey, assetKind, scenario, soc, freqBreachSec }) {
    const totalSPs = spHistory.length;
    const accepted = spHistory.filter(h => h.accepted);
    const totalAccepted = accepted.length;
    const totalRevenue = cash; // Using cash directly as it now includes settled DA
    const bestSingleSP = accepted.length ? Math.max(...accepted.map(h => h.revenue)) : 0;

    let streak = 0;
    for (let i = spHistory.length - 1; i >= 0; i--) {
        if (spHistory[i].accepted) streak++;
        else break;
    }

    // Dynamic limit check based on shared constants
    const hitSoCLimit = soc <= (MIN_SOC + 1) || soc >= (MAX_SOC - 1);

    const hadBuySellFlip = spHistory.length >= 3 &&
        spHistory.slice(-3).some(h => h.accepted && !h.isShort) &&
        spHistory.slice(-3).some(h => h.accepted && h.isShort);

    return {
        totalSPs, totalAccepted, totalRevenue, bestSingleSP, streak,
        assetKey, assetKind, scenario, daCash, soc,
        hitSoCLimit, hadBuySellFlip,
        hadFreqBreach: freqBreachSec > 0,
    };
}

export function checkAchievements(stats, alreadyEarned) {
    const newlyEarned = [];
    for (const a of ACHIEVEMENTS) {
        if (alreadyEarned.includes(a.id)) continue;
        try { if (a.check(stats)) newlyEarned.push(a); } catch { }
    }
    return newlyEarned;
}
